dwave/optimization/src/nodes/sorting.cpp

```diff
@@ -0,0 +1,119 @@
+
+/// ArgSortNode
+
+struct ArgSortNodeDataHelper_ {
+    ArgSortNodeDataHelper_(std::vector<double> values) {
+        for (ssize_t index = 0, stop = values.size(); index < stop; index++) {
+            order.emplace(values[index], index);
+        }
+
+        for (const auto& [_, index] : order) {
+            indices.push_back(index);
+        }
+    }
+
+    std::vector<double> indices;
+    std::set<std::pair<double, ssize_t>> order;
+};
+
+struct ArgSortNodeData : public ArrayNodeStateData {
+    ArgSortNodeData(std::vector<double> values)
+            : ArgSortNodeData(ArgSortNodeDataHelper_(std::move(values))) {}
+    ArgSortNodeData(ArgSortNodeDataHelper_&& helper)
+            : ArrayNodeStateData(std::move(helper.indices)), order(std::move(helper.order)) {}
+
+    /// First is the value in the original array, second is the index of the value
+    std::set<std::pair<double, ssize_t>> order;
+};
+
+ArgSortNode::ArgSortNode(ArrayNode* arr_ptr)
+        : ArrayOutputMixin(arr_ptr->shape()), arr_ptr_(arr_ptr) {
+    add_predecessor(arr_ptr);
+}
+
+double const* ArgSortNode::buff(const State& state) const {
+    return data_ptr<ArgSortNodeData>(state)->buff();
+}
+
+void ArgSortNode::commit(State& state) const { data_ptr<ArgSortNodeData>(state)->commit(); }
+
+std::span<const Update> ArgSortNode::diff(const State& state) const {
+    return data_ptr<ArgSortNodeData>(state)->diff();
+}
+
+void ArgSortNode::initialize_state(State& state) const {
+    const Array::View arr = arr_ptr_->view(state);
+
+    std::vector<double> vals{arr.begin(), arr.end()};
+    emplace_data_ptr<ArgSortNodeData>(state, vals);
+}
+
+bool ArgSortNode::integral() const { return arr_ptr_->integral(); }
+
+std::pair<double, double> ArgSortNode::minmax(
+        optional_cache_type<std::pair<double, double>> cache) const {
+    return memoize(cache, [&]() {
+        return std::make_pair(0.0, static_cast<double>(arr_ptr_->sizeinfo().max.value_or(
+                                           std::numeric_limits<ssize_t>::max())));
+    });
+}
+
+void ArgSortNode::propagate(State& state) const {
+    auto node_data = data_ptr<ArgSortNodeData>(state);
+
+    // Make the modifications to the std::set based on the updates.
+    for (const Update& update : arr_ptr_->diff(state)) {
+        if (!update.placed()) {
+            node_data->order.erase(std::make_pair(update.old, update.index));
+        }
+        if (!update.removed()) {
+            node_data->order.insert(std::make_pair(update.value, update.index));
+        }
+    }
+
+    // Assign the new order as determined by the ordering of the std::set.
+    // A further optimization could be to track the earliest modified (final) index
+    // and only assign from there which might help when all the updates only affect
+    // the end region of the final ordering.
+    node_data->assign(node_data->order |
+                      std::views::transform([](std::pair<double, ssize_t> p) { return p.second; }));
```

Comment:
I think you can do `std::views::transform(std::get<1>)` which might make let it fit on one line. Though perhaps less readable

Reply:
I tried that first, unfortunately it doesn't work with `std::transform` without annoying fiddling so I just went with what I thought was the clearer lambda. [See this SO post](https://stackoverflow.com/a/32000899)